### PR TITLE
Refine feature elimination with SHAP

### PR DESCRIPTION
**Pull request by Matgrb**
The idea is to refine the shap elimination by:
- Making it more aligned with sklearn api
- Not apply randomized search by default, the user can provide RandomizedSearchCV object or just a simple model
- Apply CV to estimate shap feature importance.

### CODE REVIEW

File: probatus/utils/__init__.py [hunk 1]
```diff
@@ -30,4 +30,5 @@
 __all__ = ['NotFittedError', 'DimensionalityError', 'UnsupportedModelError', 'Scorer', 'assure_numpy_array', 'assure_pandas_df', 'check_1d',
            'ApproximationWarning', 'class_name_from_object', 'get_scorers', 'assure_list_of_strings',
            'assure_list_values_allowed', 'warn_if_missing','TreePathFinder', 'check_numeric_dtypes',
-           'plot_distributions_of_feature', 'assure_column_names_consistency', 'shap_calc', 'shap_to_df', 'calculate_shap_importance']
+           'plot_distributions_of_feature', 'assure_column_names_consistency', 'shap_calc', 'shap_to_df', 'calculate_shap_importance',
```
Comment (timvink) [inline probatus/utils/__init__.py:33]: Minor detail, but just wondering if you ready want to expose all internal functions to users?

File: docs/tutorials/nb_shap_feature_elimination.ipynb [hunk 2]
```diff
@@ -10,17 +10,19 @@
...
+    "hyperparameters of the model at each round, to tune the model for each features set. Lastly, it supports categorical features (`object` and `category` dtype) and missing values in the data, as long as the model supports them.\n",
     " \n",
     "The main advantages of using this routine are:\n",
-    "- High computation speed compared compared to o.\n",
+    "\n",
+    "- The approach uses a tree-based model to detect the complex relations between features and the target.\n",
     "- Uses SHAP importance, which is one of the most reliable ways to estimate features importance. Unlike many other techniques, it works with missing values and categorical variables.\n",
-    "- Uses Cross-validation to optimize hyperparameters for each features set. This way you can assess if removal of a given feature reduces the predictive power, or simply requires additional tuning of the model.\n",
+    "- Allows to use GridsearchCV and RandomizedSearchCV in order to optimize hyperparameters at each iteration. This way you can assess if the removal of a given feature reduces the predictive power, or simply requires additional tuning of the model.\n",
```
Comment (timvink) [inline docs/tutorials/nb_shap_feature_elimination.ipynb:25]: Add links to sklearn classes

File: docs/tutorials/nb_shap_feature_elimination.ipynb [hunk 3]
```diff
@@ -293,14 +294,15 @@
    "source": [
     "## Set up the model and model tuning\n",
     "\n",
-    "You need to set up the model and it's hyperparameter optimization space/grid for the Backwards Feature Elimination. `Probatus` requires a **tree-based binary classifier** in order to speed up the computation of SHAP feature importance at each step. \n",
+    "You need to set up the model that you would like to use in the feature elimination. `Probatus` requires a **tree-based binary classifier** in order to speed up the computation of SHAP feature importance at each step. \n",
```
Comment (timvink) [inline docs/tutorials/nb_shap_feature_elimination.ipynb:297]: Minor detail, but I see you capitalize probatus. Is it `Probatus` or `probatus`? I recommend lower-case.

Comment (timvink) [reply docs/tutorials/nb_shap_feature_elimination.ipynb:297]: Another thought. If you use `KernelExplainer` instead of `TreeExplainer`, you get support for any model. Perhaps in a future version we could let the use specify which they would like to use. I would rephrase "probatus requires a tree-based.."  to something like "probatus currently only supports tree-based binary classifiers"

File: probatus/feature_elimination/feature_elimination.py [hunk 4]
```diff
@@ -1,26 +1,32 @@
...
-     features set, starting from all available features, a model is optimized (e.g. using RandomSearchCV) and trained.
-     At the end of each round, the n lowest SHAP feature importance features are removed and the model results are
-     stored. The user can plot the performance of the model for each round, and select the optimal number of features
-     and the features set.
-
-    The functionality is similar to [sklearn.feature_selection.RFECV](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFECV.html),
-     yet, it removes the lowest importance features based on SHAP features importance and optimizes the hyperparameters
-     of the model at each round.
+    This class performs Backwards Recursive Feature Elimination, using SHAP feature importance. At each round, for a
+        given feature set, starting from all available features, Cross-Validation in order to estimate the SHAP feature
```
Comment (timvink) [inline probatus/feature_elimination/feature_elimination.py:15]: This sentence is broken "[...] from all availabe features, cross-validation in order [..]"

File: probatus/feature_elimination/feature_elimination.py [hunk 5]
```diff
@@ -1,26 +1,32 @@
...
+        importance. For each CV iteration, the model is fitted on n-1 folds, and the SHAP values are computed for the
+        last fold. Later, the SHAP values are combined for all test folds, and the `step` lowest SHAP feature importance
```
Comment (timvink) [inline probatus/feature_elimination/feature_elimination.py:17]: Text might be clear if you use bullet points to describe the steps / loops.

File: probatus/feature_elimination/feature_elimination.py [hunk 6]
```diff
@@ -1,26 +1,32 @@
...
+        features are removed. At the end of the process, the user can plot the performance of the model for each
+        iteration, and select the optimal number of features and the features set.
+
+    The functionality is similar to sklearn.feature_selection.RFECV, yet, it removes the lowest importance features
+        based on SHAP features importance. It also supports the use of GridSearchCV and RandomizedSearchCV passed as a
+        clf, thanks to which you can perform hyperparameter optimization at each step of the search. hyperparameters of
```
Comment (timvink) [inline probatus/feature_elimination/feature_elimination.py:23]: broken sentence. Also, I think the intro "this class is similar to sklearn RFECV" is useful and might be at the top of the docstring.

File: probatus/feature_elimination/feature_elimination.py [hunk 7]
```diff
@@ -1,26 +1,32 @@
...
-     yet, it removes the lowest importance features based on SHAP features importance and optimizes the hyperparameters
-     of the model at each round.
+    This class performs Backwards Recursive Feature Elimination, using SHAP feature importance. At each round, for a
+        given feature set, starting from all available features, Cross-Validation in order to estimate the SHAP feature
+        importance. For each CV iteration, the model is fitted on n-1 folds, and the SHAP values are computed for the
+        last fold. Later, the SHAP values are combined for all test folds, and the `step` lowest SHAP feature importance
+        features are removed. At the end of the process, the user can plot the performance of the model for each
+        iteration, and select the optimal number of features and the features set.
+
+    The functionality is similar to sklearn.feature_selection.RFECV, yet, it removes the lowest importance features
```
Comment (timvink) [inline probatus/feature_elimination/feature_elimination.py:21]: add link to sklearn module

File: probatus/feature_elimination/feature_elimination.py [hunk 8]
```diff
@@ -1,26 +1,32 @@
...
+        based on SHAP features importance. It also supports the use of GridSearchCV and RandomizedSearchCV passed as a
+        clf, thanks to which you can perform hyperparameter optimization at each step of the search. hyperparameters of
+        the model at each round, to tune the model for each features set. Lastly, it supports categorical features
+        (object and category dtype) and missing values in the data, as long as the model supports them.
 
     We recommend using LightGBM model, because by default it handles missing values and categorical features. In case
```
Comment (timvink) [inline probatus/feature_elimination/feature_elimination.py:27]: "a LightGBM model" , and also add link

File: probatus/feature_elimination/feature_elimination.py [hunk 9]
```diff
@@ -72,48 +80,37 @@ def __init__(self, clf, search_space, search_schema='random', step=1, min_featur
...
-                of `random` search_schema, then this value is passed to RandomSearchCV as `param_distributions` parameter.
+            step (Optional, int or float): Number of lowest importance features removed each round. If it
+                 is an int, then each round such number of features is discarded. If float, such percentage of remaining
+                 features (rounded down) is removed each iteration. It is recommended to use float, since it is faster
+                 for a large number of features, and slows down and becomes more precise towards less features. Note:
+                 the last round may remove fewer features in order to reach min_features_to_select.
 
-            search_schema (Optional, str): The hyperparameter search algorithm that should be used to optimize the model.
-                It can be one of the following:
+            min_features_to_select (Optional, unt): Minimum number of features to be kept. This is a stopping criterion
```
Comment (timvink) [inline probatus/feature_elimination/feature_elimination.py:89]: int not `unt`

File: probatus/feature_elimination/feature_elimination.py [hunk 10]
```diff
@@ -72,48 +80,37 @@ def __init__(self, clf, search_space, search_schema='random', step=1, min_featur
...
+                of the feature elimination. By default the process stops when one feature is left.
 
-                - `random`: [RandomSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html)
-                  which randomly selects hyperparameters from the prowided param_grid, and performs optimization using
-                  Cross-Validation. It is recommended option, when you optimize a large number of hyperparameters.
-                - `grid`: [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)
-                  which searches through all permutations of hyperparameters values from param_grid, and performs
-                  optimization using Cross-Validation. It is recommended option, for low number of hyperparameters.
+            cv (Optional, int, cross-validation generator or an iterable): Determines the cross-validation splitting
```
Comment (timvink) [inline probatus/feature_elimination/feature_elimination.py:92]: You mention 'cross validation generator' but if I understand correctly, you got smart and also allow for 'cross validation estimators'. See https://scikit-learn.org/stable/glossary.html#term-cv-splitter for terminology. I recommend using the same terms & mentioning here that both the CV generators and estimators are allowed

**Review by timvink — COMMENTED** [review]
Nice improvement!

File: probatus/utils/__init__.py (hunk 1 above)
Comment (Matgrb) [reply probatus/utils/__init__.py:33]: I think they are already exposed that they can import them, the assumption is if they are not described in the API they will not be used so much for now. In the future we can describe them in the docs.

File: docs/tutorials/nb_shap_feature_elimination.ipynb (hunk 2 above)
Comment (Matgrb) [reply docs/tutorials/nb_shap_feature_elimination.ipynb]: Done

File: docs/tutorials/nb_shap_feature_elimination.ipynb (hunk 3 above)
Comment (Matgrb) [reply docs/tutorials/nb_shap_feature_elimination.ipynb]: Done, used lower-case

File: probatus/feature_elimination/feature_elimination.py (hunk 6 above)
Comment (Matgrb) [reply probatus/feature_elimination/feature_elimination.py]: I would leave it below the description on what it does.

File: probatus/feature_elimination/feature_elimination.py (hunk 10 above)
Comment (Matgrb) [reply probatus/feature_elimination/feature_elimination.py]: In this case not. It supports some specific cross validation estimators (GridsearchCV and RandomsearchCV) as clf parameter,  but not as cv parameter.

Indeed, I added them to the clf parameter description. Nice catch!
